src/components/TransformWorkflow/FlowEditor/Components/OperationPanel/Forms/UnionTablesOpForm.tsx

```diff
@@ -94,29 +96,85 @@ const UnionTablesOpForm = ({
     }
   };
 
-  const addDummyNodeAndEdge = (model: DbtSourceModel) => {
-    const dummySourceNodeData: any = {
-      id: model.id,
-      type: SRC_MODEL_NODE,
-      data: model,
-      position: {
-        x: node ? node?.xPos + 150 : 100,
-        y: node?.yPos,
-      },
-    };
-    const newEdge: any = {
-      id: `${dummySourceNodeData.id}_${dummyNodeId}`,
-      source: dummySourceNodeData.id,
-      target: dummyNodeId,
-      sourceHandle: null,
-      targetHandle: null,
-    };
-    addNodes([dummySourceNodeData]);
-    addEdges([newEdge]);
-  };
+  const clearAndAddDummyModelNode = (
+    model: DbtSourceModel | undefined | null,
+    index: number
+  ) => {
+    index = index - 1;
+    let currentModelDummyNodeIds = modelDummyNodeIds.current;
+    const edges: Edge[] = getEdges();
+    console.log('mnode', model);
+    console.log('index', index);
+
+    // make sure the arry is of correct length
+    if (currentModelDummyNodeIds.length < index + 1) {
+      currentModelDummyNodeIds = currentModelDummyNodeIds.concat(
+        Array(index + 1 - currentModelDummyNodeIds.length).fill(null)
+      );
+    }
+
+    // remove dummy node if present
+    if (currentModelDummyNodeIds[index]) {
+      if (
+        edges.filter(
+          (edge: Edge) =>
+            edge.source === currentModelDummyNodeIds[index] &&
+            edge.target === dummyNodeId
+        ).length <= 1
+      )
+        deleteElements({ nodes: [{ id: currentModelDummyNodeIds[index] }] });
+      else {
+        // if removeNodeId has multiple edges, the remove the dummy one we just created
+        let removeEdges: Edge[] = edges.filter(
+          (edge: Edge) =>
+            edge.source === currentModelDummyNodeIds[index] &&
+            edge.target === dummyNodeId
+        );
+        deleteElements({
+          edges: removeEdges.map((edge: Edge) => ({ id: edge.id })),
+        });
+      }
+      currentModelDummyNodeIds[index] = null;
+    }
+
+    // create and update the new dummy node at same index
+    if (model) {
+      let dummySourceNodeData: any = getNodes().find(
+        (node) => node.id === model.id
+      );
+
+      if (!dummySourceNodeData) {
+        // create a new dummy node if its not on the canvas
+        const { x: xnew, y: ynew } = getNextNodePosition([
+          {
+            position: { x: node?.xPos, y: node?.yPos },
+            height: 400 * (index + 1),
+          },
+        ]);
+
+        dummySourceNodeData = {
+          id: model.id,
```

Comment:
This seems common in the function clearAndAddDummyModelNode. Can we extract this logic out and reuse?